plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol

```diff
@@ -0,0 +1,119 @@
+    }
+
+    function isCorrectStateTransition(
+        bytes calldata inFlightTx,
+        bytes[] calldata inputTxs,
+        uint256[] calldata inputUtxosPos
+    )
+        external
+        pure
+        returns (bool)
+    {
+        require(inputTxs.length == inputUtxosPos.length, "Input transactions do not match input utxos positions");
+
+        //TODO: refactor that to smaller function as soon as this issue is resolved: https://github.com/ethereum/solidity/issues/6835
+        WireTransaction.Output[] memory inputs = new WireTransaction.Output[](inputTxs.length);
+        for (uint i = 0; i < inputTxs.length; i++) {
+            uint16 outputIndex = UtxoPosLib.outputIndex(UtxoPosLib.UtxoPos(inputUtxosPos[i]));
+            WireTransaction.Output memory output = WireTransaction.getOutput(inputTxs[i], outputIndex);
+            inputs[i] = output;
+        }
+
+        WireTransaction.Output[] memory outputs = new WireTransaction.Output[](inputTxs.length);
+        PaymentTransactionModel.Transaction memory transaction = PaymentTransactionModel.decode(inFlightTx);
+        for (uint i = 0; i < transaction.outputs.length; i++) {
+            outputs[i] = WireTransaction.Output(transaction.outputs[i].amount, transaction.outputs[i].outputGuard, transaction.outputs[i].token);
+        }
+
+        return _isCorrectStateTransition(inputs, outputs);
+    }
+
+    function _isCorrectStateTransition(
+        WireTransaction.Output[] memory inputs,
+        WireTransaction.Output[] memory outputs
+    )
+        private
+        pure
+        returns (bool)
+    {
+        bool correctTransition = true;
+        uint i = 0;
+        while (correctTransition && i < outputs.length) {
+            address token = outputs[i].token;
+            WireTransaction.Output[] memory inputsForToken = filterWithToken(inputs, token);
+            WireTransaction.Output[] memory outputsForToken = filterWithToken(outputs, token);
+
+            correctTransition = isCorrectSpend(inputsForToken, outputsForToken);
+            i += 1;
+        }
+        return correctTransition;
+    }
+
+    function filterWithToken(
+        WireTransaction.Output[] memory outputs,
+        address token
+    )
+        private
+        pure
+        returns (WireTransaction.Output[] memory)
+    {
+        // it is needed to calculate the size of the filtered array
+        uint256 arraySize = 0;
+        for (uint i = 0; i < outputs.length; ++i) {
+            if (outputs[i].token == token) {
+                arraySize += 1;
+            }
+        }
+
+        WireTransaction.Output[] memory outputsWithToken = new WireTransaction.Output[](arraySize);
+        uint j = 0;
+        for (uint i = 0; i < outputs.length; ++i) {
+            if (outputs[i].token == token) {
+                outputsWithToken[j] = outputs[i];
+                j += 1;
+            }
+        }
+
+        return outputsWithToken;
+    }
+
+    function isCorrectSpend(
```

Comment:
Minor: would `isCorrectSumRelationship` better describe the function?

Reply:
done

Reply:
done

Reply:
haha not really done XD but this is minor anyway